docs/debugging.md

```diff
@@ -0,0 +1,179 @@
+# Uncovering concurrency issues, testing and debugging
+Until now, the GIL has allowed developers to elide and get away from
+concurrency issues when writing parallel programs, since the GIL ensured that
+all thread execution got serialized when any of them tried to simultaneously
+write or read a variable defined in the interpreter. Under the new free-threaded
+paradigm, developers now must think about concurrency, distribution and
+parallelism constructs that will allow them to exploit the maximum performance
+of their parallel programs.
+
+Usually, concurrency issues arise when two or more threads try to modify the
+same value in memory. In Python, this commonly occurs when a class or function
+defines shared state, either via an attribute or a variable that can modified
+from each thread execution scope.
```

Comment:
Probably worth adding a note about static state in C or Cython extensions after this sentence. Also note that shared python state is "safe" in the sense that concurrent access cannot generate C thread safety issues. Static C state that assumed the GIL prevented concurrent access does not have that guarantee.